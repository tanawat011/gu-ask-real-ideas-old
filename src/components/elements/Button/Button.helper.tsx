import clsx from 'clsx'

import { BUTTON_STYLE } from '@constants'
import { PropButton } from '.'

export const renderPrimaryClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-primary-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-primary-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-primary-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-primary-default text-primary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-primary-default text-primary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-primary-default-flat text-primary-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-primary-default', classHover())
  }

  return clsx(
    'bg-primary-default text-white',
    classHover(),
    !isLoading && 'active:bg-primary-pressed',
    hasShadow && 'shadow-nextui shadow-primary-default',
  )
}

export const renderSecondaryClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-secondary-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-secondary-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-secondary-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-secondary-default text-secondary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-secondary-default text-secondary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx(
      'bg-secondary-default-flat text-secondary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-secondary-default', classHover())
  }

  return clsx(
    'bg-secondary-default text-white',
    classHover(),
    !isLoading && 'active:bg-secondary-pressed',
    hasShadow && 'shadow-nextui shadow-secondary-default',
  )
}

export const renderSuccessClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-success-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-success-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-success-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-success-default text-success-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-success-default text-success-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-success-default-flat text-success-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-success-default', classHover())
  }

  return clsx(
    'bg-success-default text-white',
    classHover(),
    !isLoading && 'active:bg-success-pressed',
    hasShadow && 'shadow-nextui shadow-success-default',
  )
}

export const renderWarningClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-warning-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-warning-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-warning-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-warning-default text-warning-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-warning-default text-warning-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-warning-default-flat text-warning-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-warning-default', classHover())
  }

  return clsx(
    'bg-warning-default text-white',
    classHover(),
    !isLoading && 'active:bg-warning-pressed',
    hasShadow && 'shadow-nextui shadow-warning-default',
  )
}

export const renderInfoClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-info-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-info-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-info-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-info-default text-info-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-info-default text-info-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-info-default-flat text-info-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-info-default', classHover())
  }

  return clsx(
    'bg-info-default text-white',
    classHover(),
    !isLoading && 'active:bg-info-pressed',
    hasShadow && 'shadow-nextui shadow-info-default',
  )
}

export const renderDangerClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-danger-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-danger-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-danger-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-danger-default text-danger-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-danger-default text-danger-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-danger-default-flat text-danger-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-danger-default', classHover())
  }

  return clsx(
    'bg-danger-default text-white',
    classHover(),
    !isLoading && 'active:bg-danger-pressed',
    hasShadow && 'shadow-nextui shadow-danger-default',
  )
}
