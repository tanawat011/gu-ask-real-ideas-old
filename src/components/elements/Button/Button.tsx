import { MouseEvent, cloneElement } from 'react'

import {
  BUTTON_STYLE,
  VARIANT,
  LOADING_COLOR,
  LOADING_STYLE,
  SIZE,
} from '@constants'
import { Loading } from '@element/Loader'
import { RegEx } from '@helpers'
import { clsx } from '@libs'

export type PropButton = {
  width?: number
  label?: string
  size?: SIZE
  variant?: VARIANT
  gradients?: [string, string, string] | [string, string] | [string]
  btnStyle?: BUTTON_STYLE
  icon?: React.ReactElement
  iconRight?: React.ReactElement
  iconFit?: boolean
  disabled?: boolean
  loading?: boolean
  rounded?: boolean
  rippleEffect?: boolean
  wAuto?: boolean
  shadow?: boolean
  onClick?: (event: MouseEvent) => void
}

export const Button: React.FC<PropButton> = (props) => {
  const {
    width,
    label,
    size = SIZE.MD,
    variant = VARIANT.PRIMARY,
    gradients,
    icon,
    iconRight,
    btnStyle = BUTTON_STYLE.DEFAULT,
    iconFit = false,
    disabled = false,
    loading = false,
    rounded = false,
    rippleEffect = true,
    wAuto = false,
    shadow,
    onClick,
  } = props

  const createRippleEffect = (event: MouseEvent) => {
    const button = event.currentTarget as EventTarget &
      Element & { offsetLeft: number; offsetTop: number }

    const circle = document.createElement('span')
    const diameter = Math.max(button.clientWidth, button.clientHeight)
    const radius = diameter / 2

    circle.style.width = circle.style.height = `${diameter}px`
    circle.style.left = `${event.clientX - button.offsetLeft - radius}px`
    circle.style.top = `${event.clientY - button.offsetTop - radius}px`
    circle.classList.add('ripple')

    if (btnStyle !== BUTTON_STYLE.DEFAULT) {
      circle.classList.add(`btn__bg-${variant.toLowerCase()}`)
    } else {
      circle.classList.add('bg-black')
    }

    const ripple = button.getElementsByClassName('ripple')[0]

    if (ripple) {
      ripple.remove()
    }

    button.appendChild(circle)
  }

  const isDisabled = () => disabled || loading

  const renderCursor = () => (disabled || loading) && 'cursor-not-allowed'
  const renderDisabled = () => (disabled || loading) && 'btn__disabled'
  const renderRounded = () => rounded && 'btn__rounded'
  const renderAutoWidth = () => wAuto && 'btn__auto-w'
  const renderSize = () => `btn__${size.toLowerCase()}`

  const renderVariant = () => {
    const _variant = variant.toLowerCase()
    const _style = btnStyle.toLowerCase()

    if (disabled) {
      if (btnStyle === BUTTON_STYLE.DEFAULT) {
        return `btn__${_variant}-disabled`
      }

      return `btn__${_variant}-${_style}-disabled`
    }

    if (gradients) {
      const gFirst = `from-${
        RegEx.colorHex.test(gradients[0]) ? `[${gradients[0]}]` : gradients[0]
      }`
      let gSecond = ''
      let gThird = ''

      if (gradients.length === 2) {
        gSecond = `to-${
          RegEx.colorHex.test(gradients[1]) ? `[${gradients[1]}]` : gradients[1]
        }`
      }

      if (gradients.length === 3) {
        gSecond = `via-${
          RegEx.colorHex.test(gradients[1]) ? `[${gradients[1]}]` : gradients[1]
        }`
        gThird = `to-${
          RegEx.colorHex.test(gradients[2]) ? `[${gradients[2]}]` : gradients[2]
        }`
      }

      return clsx('bg-gradient-to-r', gFirst, gSecond, gThird)
    }

    if (btnStyle === BUTTON_STYLE.DEFAULT) {
      return clsx(
        `btn__${_variant}`,
        loading && `btn__${_variant}-loading`,
        shadow && `btn__${_variant}-shadow`,
      )
    }

    return clsx(
      `btn__${_variant}-${_style}`,
      loading && `btn__${_variant}-${_style}-loading`,
    )
  }

  const renderIcon = (
    _icon: React.ReactElement,
    position: 'left' | 'right',
  ) => {
    const isLeft = position === 'left'

    let classPaddingIcon = ''
    let classWidthIcon = label ? 'w-8' : 'w-6'

    if (size === SIZE.XS) {
      classWidthIcon = label ? 'w-6' : 'w-4'
    }

    if (label) {
      classPaddingIcon = isLeft ? 'pr-2' : 'pl-2'
    }

    const classIcon = clsx(
      classWidthIcon,
      classPaddingIcon,
      iconFit && 'absolute',
      iconFit && isLeft ? 'left-4' : 'right-4',
    )

    return cloneElement(_icon, { className: classIcon })
  }

  const renderChild = () => {
    return loading ? (
      <Loading loaderStyle={LOADING_STYLE.DASH} color={LOADING_COLOR.WHITE} />
    ) : (
      <>
        {icon && renderIcon(icon, 'left')}
        {label}
        {iconRight && renderIcon(iconRight, 'right')}
      </>
    )
  }

  return (
    <button
      className={clsx(
        'btn',
        renderCursor(),
        renderDisabled(),
        renderRounded(),
        renderAutoWidth(),
        renderSize(),
        renderVariant(),
      )}
      style={{ ...(width && { width }) }}
      disabled={isDisabled()}
      onClick={(event) => {
        if (rippleEffect) {
          createRippleEffect(event)
        }

        if (onClick) {
          onClick(event)
        }
      }}
    >
      {renderChild()}
    </button>
  )
}
