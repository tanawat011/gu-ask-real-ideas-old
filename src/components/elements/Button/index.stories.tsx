import { ComponentMeta, ComponentStory } from '@storybook/react'

import { Button } from '.'

export default {
  title: 'Button/Default',
  component: Button,
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Default = Template.bind({})

Default.args = {
  title: 'Button',
  size: 'medium',
  color: 'primary-default',
}
