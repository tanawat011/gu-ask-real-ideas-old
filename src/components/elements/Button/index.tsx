import clsx from 'clsx'
import { VFC } from 'react'
import Image from 'next/image'

interface ButtonType {
  title: string
  size: 'small' | 'medium' | 'large'
  color: 'yellow' | 'red' | 'white' | 'bg'
  icon?: string
  isDisabled?: boolean
  width?: number
}

export const Button: VFC<ButtonType> = ({
  width,
  title,
  size,
  color,
  icon,
  isDisabled = false,
}) => {
  const renderSize = (size: string) => {
    if (size === 'small') {
      return 'h-6-5 w-52-7 text-xss rounded'
    }

    if (size === 'small' && color === 'white') {
      return 'h-6 w-28-1 text-xss rounded'
    }

    if (size === 'medium') {
      return 'h-7-5 w-64-2 text-xss rounded-md'
    }

    if (size === 'medium' && color === 'white') {
      return 'h-6 w-64-8 text-xss rounded-md'
    }

    if (size === 'large') {
      return 'h-10 w-64-2 text-xs rounded-lg'
    }

    if (size === 'large' && color === 'white') {
      return 'h-6 w-64-8 text-xs rounded-lg'
    }
  }

  const renderColor = (color: string) => {
    if (color === 'yellow') {
      return clsx(
        'bg-action-default hover:bg-action-hover active:bg-action-pressed',
        isDisabled && 'disabled:bg-action-default opacity-40',
      )
    }

    if (color === 'red') {
      return clsx(
        'bg-secondary-candy hover:bg-candy-hover active:bg-candy-pressed',
        isDisabled && 'disabled:bg-secondary-candy opacity-40',
      )
    }

    if (color === 'white') {
      return clsx(
        'bg-neutral-fade-100 hover:bg-neutral-gray-200 active:bg-neutral-dark-400',
        isDisabled && 'disabled:bg-neutral-fade-100 opacity-40',
      )
    }

    if (color === 'bg') {
      return clsx(
        'bg-neutral-gray-300 hover:bg-neutral-gray-200 active:bg-neutral-dark-400',
        isDisabled && 'disabled:bg-neutral-gray-300 opacity-40',
      )
    }
  }

  return (
    <button
      className={`
        ${renderSize(size)}
        ${renderColor(color)}
        flex items-center justify-center font-medium text-neutral-black
      `}
      style={{ width }}
      disabled={isDisabled}
    >
      {icon && (
        <Image className='w-3 h-3 m-1 pointer-events-none' src={icon} alt='' />
      )}{' '}
      {title}
    </button>
  )
}
