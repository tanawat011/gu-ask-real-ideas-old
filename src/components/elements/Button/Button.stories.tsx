import { ComponentMeta, ComponentStory } from '@storybook/react'

import { Button } from '@element/Button/Button'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    label: {
      name: 'Label',
      table: {
        category: 'Text',
      },
    },
    size: {
      name: 'Size',
      table: {
        category: 'Size',
      },
    },
    color: {
      name: 'Color',
      table: {
        category: 'Colors',
      },
    },
    isLoading: {
      name: 'Set loading',
      table: {
        category: 'Setting',
      },
    },
    isDisabled: {
      name: 'Set disabled',
      table: {
        category: 'Setting',
      },
    },
  },
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Primary = Template.bind({})

Primary.args = {
  label: 'Button',
  size: 'medium',
  color: 'primary',
}
Primary.parameters = {
  docs: {
    description: {
      component: 'Button primary',
    },
  },
}

export const Secondary = Template.bind({})

Secondary.args = {
  label: 'Button',
  size: 'medium',
  color: 'secondary',
}
Secondary.parameters = {
  docs: {
    description: {
      component: 'Button secondary',
    },
  },
}
