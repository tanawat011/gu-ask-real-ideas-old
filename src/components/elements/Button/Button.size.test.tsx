import { render } from '@testing-library/react'

import { Button } from '@element/Button'
import { BUTTON_STYLE, COLOR, SIZE } from '@constants'

type TestButtonSize = {
  _size: SIZE
  size: string
  _color: COLOR
  color: string
  _style: BUTTON_STYLE
  style: string
  h: string
  w: string
  px: string
  fontWeight: string
  classAssemble: string
  disabled?: boolean
  loading?: boolean
  shadow?: boolean
  rounded?: boolean
}

const baseClass =
  'relative flex justify-center items-center overflow-hidden transition-colors active:scale-95'
const baseClassDisabled =
  'relative flex justify-center items-center overflow-hidden transition-colors cursor-not-allowed'

describe('<Button />', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  let options: TestButtonSize[] = []

  Object.keys(SIZE).forEach((size) => {
    const isMedium = size.toLowerCase() === 'md'
    const optionSize = isMedium ? 'base' : size.toLowerCase()

    let h = ''
    let w = ''
    let px = ''
    let fontWeight = ''
    let classAssemble = ''

    switch (size.toLowerCase()) {
      case SIZE.XS:
        h = 'h-6'
        w = 'w-20'
        px = 'px-2'
        fontWeight = 'font-semibold'
        break

      case SIZE.SM:
        h = 'h-8'
        w = 'w-36'
        px = 'px-2'
        fontWeight = 'font-semibold'
        break

      case SIZE.LG:
        h = 'h-11'
        w = 'w-60'
        px = 'px-6'
        fontWeight = 'font-bold'
        break

      case SIZE.XL:
        h = 'h-14'
        w = 'w-72'
        px = 'px-6'
        fontWeight = 'font-bold'
        break

      case SIZE.MD:
      default:
        h = 'h-10'
        w = 'w-48'
        px = 'px-4'
        fontWeight = 'font-semibold'
        break
    }

    const baseOfBaseClass = `${baseClass} ${h} ${px} text-${optionSize} ${fontWeight}`

    Object.keys(COLOR).forEach((color) => {
      const optionColor = color.toLowerCase()

      Object.keys(BUTTON_STYLE).forEach((style) => {
        switch (style.toLowerCase()) {
          case BUTTON_STYLE.OUTLINE:
            classAssemble = `${baseOfBaseClass} rounded-xl bg-transparent border-solid border-2 border-${optionColor}-default text-${optionColor}-default hover:opacity-80`
            break

          case BUTTON_STYLE.FLAT:
            classAssemble = `${baseOfBaseClass} rounded-xl bg-${optionColor}-default-flat text-${optionColor}-default hover:opacity-80`
            break

          case BUTTON_STYLE.GHOST:
            classAssemble = `${baseOfBaseClass} rounded-xl bg-transparent border-solid border-2 border-${optionColor}-default text-${optionColor}-default hover:text-white hover:bg-${optionColor}-default`
            break

          case BUTTON_STYLE.LIGHT:
            classAssemble = `${baseOfBaseClass} rounded-xl bg-transparent text-${optionColor}-default hover:opacity-80`
            break

          case BUTTON_STYLE.DEFAULT:
          default:
            classAssemble = `${baseOfBaseClass} rounded-xl bg-${optionColor}-default text-white hover:bg-${optionColor}-hover active:bg-${optionColor}-pressed`
            break
        }

        options.push({
          _size: isMedium ? SIZE.MD : SIZE[size as keyof typeof SIZE],
          size: optionSize,
          _color: COLOR[color as keyof typeof COLOR],
          color: optionColor,
          _style: BUTTON_STYLE[style as keyof typeof BUTTON_STYLE],
          style: style.toLowerCase(),
          h,
          w,
          px,
          fontWeight,
          classAssemble,
        })
      })
    })
  })

  options = [
    ...options,
    ...options.map((o) => {
      let classAssemble = `${baseClassDisabled} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl disabled:bg-${o.color}-disabled text-nextui-gray-600`

      if (o._style === BUTTON_STYLE.LIGHT) {
        classAssemble = `${baseClassDisabled} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl disabled:bg-transparent text-nextui-gray-600`
      }

      return {
        ...o,
        classAssemble,
        disabled: true,
      }
    }),
    ...options.map((o) => {
      let classAssemble = `${baseClassDisabled} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl bg-${o.color}-default text-white`

      if (
        o._style === BUTTON_STYLE.GHOST ||
        o._style === BUTTON_STYLE.OUTLINE
      ) {
        classAssemble = `${baseClassDisabled} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl bg-transparent border-solid border-2 border-${o.color}-default text-${o.color}-default`
      }

      if (o._style === BUTTON_STYLE.FLAT) {
        classAssemble = `${baseClassDisabled} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl bg-${o.color}-default-flat text-${o.color}-default`
      }

      if (o._style === BUTTON_STYLE.LIGHT) {
        classAssemble = `${baseClassDisabled} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl bg-transparent text-${o.color}-default`
      }

      return {
        ...o,
        classAssemble,
        loading: true,
      }
    }),
    ...options.map((o) => {
      let classAssemble = `${baseClass} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl bg-${o.color}-default text-white hover:bg-${o.color}-hover active:bg-${o.color}-pressed shadow-nextui shadow-${o.color}-default`

      if (
        o._style === BUTTON_STYLE.GHOST ||
        o._style === BUTTON_STYLE.OUTLINE
      ) {
        classAssemble = `${baseClass} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl bg-transparent border-solid border-2 border-${o.color}-default text-${o.color}-default`
      }

      if (o._style === BUTTON_STYLE.FLAT) {
        classAssemble = `${baseClass} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl bg-${o.color}-default-flat text-${o.color}-default`
      }

      if (o._style === BUTTON_STYLE.LIGHT) {
        classAssemble = `${baseClass} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} rounded-xl bg-transparent text-${o.color}-default hover:opacity-80`
      }

      return {
        ...o,
        classAssemble,
        shadow: true,
      }
    }),
    ...options.map((o) => {
      let classAssemble = `${baseClass} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} bg-${o.color}-default text-white hover:bg-${o.color}-hover active:bg-${o.color}-pressed`

      if (o._style === BUTTON_STYLE.OUTLINE) {
        classAssemble = `${baseClass} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} bg-transparent border-solid border-2 border-${o.color}-default text-${o.color}-default hover:opacity-80`
      }

      if (o._style === BUTTON_STYLE.GHOST) {
        classAssemble = `${baseClass} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} bg-transparent border-solid border-2 border-${o.color}-default text-${o.color}-default hover:text-white hover:bg-${o.color}-default`
      }

      if (o._style === BUTTON_STYLE.FLAT) {
        classAssemble = `${baseClass} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} bg-${o.color}-default-flat text-${o.color}-default hover:opacity-80`
      }

      if (o._style === BUTTON_STYLE.LIGHT) {
        classAssemble = `${baseClass} ${o.h} ${o.px} text-${o.size} ${o.fontWeight} bg-transparent text-${o.color}-default hover:opacity-80`
      }

      return {
        ...o,
        classAssemble,
        rounded: true,
      }
    }),
  ]

  it('should render to match snapshot', () => {
    expect.assertions(1)

    const { asFragment } = render(
      <Button
        color={COLOR.PRIMARY}
        size={SIZE.MD}
        icon='ArchiveIcon'
        label='Button A'
        width={100}
        hasShadow
      />,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render without props', () => {
    expect.assertions(1)

    const { asFragment } = render(<Button label='Button A' />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render with onClick', async () => {
    expect.assertions(1)

    const onClick = jest.fn()

    const { findByRole } = render(<Button label='Button A' onClick={onClick} />)

    const button = await findByRole('button')

    button.click()

    expect(onClick).toHaveBeenCalledTimes(1)
  })

  options
    .filter(
      (o) => o._color === COLOR.PRIMARY && o._style === BUTTON_STYLE.DEFAULT,
    )
    .forEach((o) => {
      it(`should render with no auto width size ${o.size}, color ${o.color}`, async () => {
        expect.assertions(1)

        const { findByRole } = render(
          <Button label='Test size' isAutoWidth={false} size={o._size} />,
        )

        const button = await findByRole('button')

        expect(button).toHaveClass(o.w)
      })
    })

  options.forEach(
    ({
      _size,
      size,
      _color,
      color,
      _style,
      style,
      classAssemble,
      disabled,
      loading,
      shadow,
      rounded,
    }) => {
      if (!disabled && !loading && !shadow && !rounded) {
        it(`should be button size ${size}, color ${color}, style ${style}`, async () => {
          expect.assertions(1)

          const { findByRole } = render(
            <Button
              size={_size}
              color={_color}
              buttonStyle={_style}
              label='Test size'
            />,
          )
          const button = await findByRole('button')

          expect(button).toHaveClass(classAssemble)
        })
      }

      if (disabled) {
        it(`should be button size ${size}, color ${color}, style ${style}, disabled button`, async () => {
          expect.assertions(1)

          const { findByRole } = render(
            <Button
              size={_size}
              color={_color}
              buttonStyle={_style}
              label='Test size'
              isDisabled
            />,
          )
          const button = await findByRole('button')

          expect(button).toHaveClass(classAssemble)
        })
      }

      if (loading) {
        it(`should be button size ${size}, color ${color}, style ${style}, loading button`, async () => {
          expect.assertions(1)

          const { findByRole } = render(
            <Button
              size={_size}
              color={_color}
              buttonStyle={_style}
              label='Test size'
              isLoading
            />,
          )
          const button = await findByRole('button')

          expect(button).toHaveClass(classAssemble)
        })
      }

      if (shadow) {
        it(`should be button size ${size}, color ${color}, style ${style}, shadow button`, async () => {
          expect.assertions(1)

          const { findByRole } = render(
            <Button
              size={_size}
              color={_color}
              buttonStyle={_style}
              label='Test size'
              hasShadow
            />,
          )
          const button = await findByRole('button')

          expect(button).toHaveClass(classAssemble)
        })
      }

      if (rounded) {
        it(`should be button size ${size}, color ${color}, style ${style}, rounded button`, async () => {
          expect.assertions(1)

          const { findByRole } = render(
            <Button
              size={_size}
              color={_color}
              buttonStyle={_style}
              label='Test size'
              isRounded
            />,
          )
          const button = await findByRole('button')

          expect(button).toHaveClass(classAssemble)
        })
      }
    },
  )
})
