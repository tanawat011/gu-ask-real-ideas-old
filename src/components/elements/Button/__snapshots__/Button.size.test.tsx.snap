// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render to match snapshot 1`] = `
<DocumentFragment>
  <button
    class="relative flex justify-center items-center overflow-hidden transition-colors active:scale-95 h-10 px-4 text-base font-semibold rounded-xl bg-primary-default text-white hover:bg-primary-hover active:bg-primary-pressed shadow-nextui shadow-primary-default"
    style="width: 100px;"
  >
    <svg
      aria-hidden="true"
      class="w-8 pr-2 right-4"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 3a2 2 0 100 4h12a2 2 0 100-4H4z"
      />
      <path
        clip-rule="evenodd"
        d="M3 8h14v7a2 2 0 01-2 2H5a2 2 0 01-2-2V8zm5 3a1 1 0 011-1h2a1 1 0 110 2H9a1 1 0 01-1-1z"
        fill-rule="evenodd"
      />
    </svg>
    Button A
  </button>
</DocumentFragment>
`;

exports[`<Button /> should render without props 1`] = `
<DocumentFragment>
  <button
    class="relative flex justify-center items-center overflow-hidden transition-colors active:scale-95 h-10 px-4 text-base font-semibold rounded-xl bg-primary-default text-white hover:bg-primary-hover active:bg-primary-pressed"
  >
    Button A
  </button>
</DocumentFragment>
`;
